File: 9012.py
N = int(input())

for i in range(N):
    s = input()
    VPS = list(s)
    sum = 0
    for j in VPS:
        if j=='(':
           sum+=1
        else:
            sum -=1
        if sum<0:
            print("NO")
            break
    
    if sum>0:
        print("NO")
    elif sum==0: 
        print("YES")
                

File: Baekjoon/9012.cpp
#include<iostream>
#include<string>
#include<stack>
using namespace std;

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int T;
	cin >> T;
	for (int i = 0; i < T; i++) {
		string s;
		bool VPS = true;
		stack<char> stk;
		cin >> s;
		for (int i = 0; i < s.length(); i++) {
			if (s[i] == '(')
				stk.push(s[i]);
			else {
				if (!stk.empty()) {
					if (stk.top() == '(')
						stk.pop();
				}
				else {
					VPS = false;
					break;
				}
			}
		}

		if (VPS && stk.empty())
			cout << "YES" << '\n';
		else
			cout << "NO" << '\n';
	}
	return 0;
}
